File: ios/PrimesRN/PrimeFinder.m
//
//  PrimeFinder.m
//  PrimesRN
//
//  Created by Leif Terry on 10/1/2018.
//  Copyright © 2018 Facebook. All rights reserved.
//

#import "PrimeFinder.h"
#import <React/RCTLog.h>

@interface PrimeFinder()
    @property int lastRoot;     // estimate of square root of last prime found
    @property int primesFound;  // num primes found so far
    @property int nextNotable;  // send message for "notable" primes, i.e. 10th, 100th, 1000th etc.
    @property int batchSize;    // send message after finding this many primes
    @property NSMutableArray* primeList;    // list of primes found, excluding 2
@end

@implementation PrimeFinder

RCT_EXPORT_MODULE();

-(id) init
{
    self = [super init];
    if (self)
    {
        // add custom init here
    }
    return self;
}

+(BOOL) requiresMainQueueSetup
{
  return NO;
}

-(NSArray<NSString*>*) supportedEvents
{
    return @[@"foundPrime"];
}

-(void) setupToFind:(int) numToFind
{
    self.primeList = [[NSMutableArray alloc] initWithCapacity:numToFind];
    [self.primeList addObject:@2];
    [self.primeList addObject:@3];
    NSNumber* lastPrime = [self.primeList lastObject];
    assert((lastPrime.intValue % 2) == 1);

    self.primesFound = (int) self.primeList.count;
    self.lastRoot = sqrtf((float) lastPrime.intValue) + 1;

    self.nextNotable = 10;
    self.batchSize = 10;
}

-(BOOL) isPrime_v0:(int) n
{
    // simple but has redundant checks for non-prime odds such as 9, 15, 21 etc.
    for (int ii = 3; ii <= (self.lastRoot+1); ii += 2)
    {
        if ((n % ii) == 0)
        {
            return NO;
        }
    }
    return YES;
}

-(BOOL) isPrime_v1:(int) n
{
    // we know n is not even, skip index 0 == 2
    // start at index 1 == 3
    for (int ii = 1; ; ii ++)
    {
        NSNumber* prime = self.primeList[ii];
        if ((n % prime.intValue) == 0)
        {
            return NO;
        }
        if (prime.intValue > self.lastRoot)
        {
            return YES;
        }
    }
    return YES;
}

-(BOOL) isPrime_v2:(int) n
{
    // uses fast iteration
    // has unnecessary check for (num % 2) == 0
    for (NSNumber* prime in self.primeList)
    {
        if ((n % prime.intValue) == 0)
        {
            return NO;
        }
        if (prime.intValue > self.lastRoot)
        {
            return YES;
        }
    }
    return YES;
}


#define isPrime isPrime_v2

RCT_EXPORT_METHOD(findPrimes:(int) numToFind)
{
    RCTLog(@"Finding 1st %d primes", numToFind);

    if (numToFind > 2)
    {
        [self setupToFind:numToFind];

        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0),
                       ^{
                           NSNumber* lastPrime = [self.primeList lastObject];
                           int foundThisBatch = 0;

                           for (int ii = lastPrime.intValue + 2; ; ii += 2) // don't bother checking evens
                           {
                               // newton's method for approximating sqrt(ii)
                               self.lastRoot -= (self.lastRoot * self.lastRoot - ii) / (2 * self.lastRoot);

                               if ([self isPrime:ii])
                               {
                                   // try a couple of things to optimize isPrime
                                   // store prime in NSArray
                                   [self.primeList addObject:[NSNumber numberWithInt:ii]];

                                   self.primesFound++;
                                   foundThisBatch++;

                                   if (self.primesFound == self.nextNotable)
                                   {
                                       [self sendEventWithName:@"foundPrime" body:@{@"prime": [NSNumber numberWithInt:ii], @"numFound" : [NSNumber numberWithInt:self.primesFound], @"isNotable" : @YES}];
                                       self.batchSize = self.nextNotable;
                                       self.nextNotable *= 10; // assumes you want 10th, 100th, 1000th etc. magic number
                                       foundThisBatch = 0;
                                   }
                                   else if (foundThisBatch == self.batchSize)
                                   {
                                       // send periodic update about primes found
                                       [self sendEventWithName:@"foundPrime" body:@{@"prime": [NSNumber numberWithInt:ii], @"numFound" : [NSNumber numberWithInt:self.primesFound], @"isNotable" : @NO}];
                                       foundThisBatch = 0;
                                   }
                                   if (self.primesFound == numToFind)
                                   {
                                       break;
                                   }
                               }
                           }
                        });
    }
    // else input is invalid or trivial, TODO handle this case
}

@end

File: android/app/src/main/java/com/primesrn/PrimeFinder.java
package com.primesrn;

import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import android.util.Log;
import java.util.Map;
import java.util.ArrayList;
import java.util.HashMap;
import android.support.annotation.Nullable;

public class PrimeFinder extends ReactContextBaseJavaModule 
{
    private ReactApplicationContext reactContext;
    private int lastRoot;           // estimate of square root of last prime found
    private int primesFound;        // num primes found so far
    private int nextNotable;        // send message for "notable" primes, i.e. 10th, 100th, 1000th etc.
    private int batchSize;          // send message after finding this many primes
    private int[] primeList;    // list of primes found

    public PrimeFinder(ReactApplicationContext reactContext)
    {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public String getName()
    {
        return "PrimeFinder";
    }

    private boolean isPrime(int num)
    {
        for (int ii = 1; ii < this.primesFound ; ii++)
        {
            int prime = this.primeList[ii];
            if ((num % prime) == 0)
            {
                // num is divisible, not prime
                return false;
            }
            if (prime > this.lastRoot)
            {
                // lastRoot is >= sqrt(num)
                // so
                // num is not divisible by any primes less than sqrt(num)
                return true;
            }
        }
        return true;
    }

    @ReactMethod
    public void findPrimes(final int numToFind) 
    {
        this.primeList = new int[numToFind];
        this.primeList[0] = 2;
        this.primeList[1] = 3;
        this.primesFound = 2;
        this.lastRoot = 2; // srt(3) rounded up
    
        this.nextNotable = 10;
        this.batchSize = 10;
     
        final PrimeFinder thisRef = this;
        Thread thread = new Thread() 
        {
            @Override
            public void run() 
            {
                    int foundThisBatch = 0;
                    for (int ii = thisRef.primeList[1] + 2; ; ii += 2) // don't bother checking evens
                    {
                        thisRef.lastRoot -= (thisRef.lastRoot * thisRef.lastRoot - ii) / (2 * thisRef.lastRoot);

                        if (isPrime(ii))
                        {
                            thisRef.primeList[thisRef.primesFound] = ii;
                            thisRef.primesFound++;
                            foundThisBatch++;

                            if (thisRef.primesFound == thisRef.nextNotable)
                            {
                                WritableMap eventData = Arguments.createMap();
                                eventData.putInt("prime", ii);
                                eventData.putInt("numFound", thisRef.primesFound);
                                eventData.putBoolean("isNotable", true);
                                sendEvent("foundPrime", eventData);

                                // increase batch size and notable threshold
                                thisRef.batchSize = thisRef.nextNotable;
                                thisRef.nextNotable *= 10;
                                foundThisBatch = 0;
                            }
                            else if (foundThisBatch == thisRef.batchSize)
                            {
                                WritableMap eventData = Arguments.createMap();
                                eventData.putInt("prime", ii);
                                eventData.putInt("numFound", thisRef.primesFound);
                                eventData.putBoolean("isNotable", false);
                                sendEvent("foundPrime", eventData);

                                // reset batch, keep current size
                                foundThisBatch = 0;
                            }
                            if (thisRef.primesFound == numToFind)
                            {
                                break;
                            }
                        }
                    }
            }
        };
        thread.start();
    }

    private void sendEvent(String eventName, @Nullable WritableMap params) 
    {
      this.reactContext
        .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
        .emit(eventName, params);
    }
}